lib/LaTeXML/Engine/TeX_Box.pool.ltxml

```diff
@@ -515,85 +519,75 @@ DefRegister('\dp Number', Dimension(0),
     $stuff->setDepth($value) if $stuff;
     return; });
 
+# Adjust the default color of $box to be the current color,
+# creating a copy of the box if needed.
 sub adjustBoxColor {
   my ($box) = @_;
   my $font = LookupValue('font');
-  if (my $color = $font && $font->getColor) {
-    if (!Black->equals($color)) {
-      adjustBoxColor_rec($color, {}, $box); } }
-  return; }
+  my $color = $font && $font->getColor;
+  my $bgcolor = $font && $font->getBackground;
+  if(($color && !Black->equals($color)) || ($bgcolor && !White->equals($bgcolor))){
+    return adjustBoxColor_rec($color, $bgcolor, {}, $box); }
+  return $box; }
 
 sub adjustBoxColor_rec {
   no warnings 'recursion';
-  my ($color, $adjusted, @boxes) = @_;
-  foreach my $box (@boxes) {
-    next unless defined $box;
-    next if $$adjusted{$box};    # since we do args AND props, be careful not to adjust twice!
-    $$adjusted{$box} = 1;
-    my $r = ref $box;
-    next unless $r && ($r !~ /(?:SCALAR|HASH|ARRAY|CODE|REF|GLOB|LVALUE)/) && $r->isaBox;
-    # NASTY access to internal structure; but worth a whole API for this one hack???
-    if ($r eq 'LaTeXML::Core::Box') {
-      adjustBoxColor_internal($color, $box); }
-    elsif ($r eq 'LaTeXML::Core::List') {
-      adjustBoxColor_rec($color, $adjusted, $box->unlist); }
-    elsif ($r eq 'LaTeXML::Core::Whatsit') {
-      adjustBoxColor_internal($color, $box);
-      # now recurse on contained boxes (args AND properties!)
-      adjustBoxColor_rec($color, $adjusted, $box->getArgs);
-      adjustBoxColor_rec($color, $adjusted, values %{ $box->getPropertiesRef }); }
-    elsif ($r eq 'LaTeXML::Core::Alignment') {
-      foreach my $row (@{ $$box{rows} }) {
-        foreach my $col (@{ $$row{columns} }) {
-          adjustBoxColor_rec($color, $adjusted, $$col{boxes}->unlist) if $$col{boxes}; } } }
-  }
-  return; }
-
-sub adjustBoxColor_internal {
-  my ($color, $box) = @_;
-  if (my $font = $box->getFont) {
-    $box->setFont($font->merge(color => $color)); }
-  return; }
+  my ($fg, $bg, $cache, $box) = @_;
+  return $box unless $box;
+  return $$cache{$box} if $$cache{$box};    # Nested, possibly circular, structures
+  my $type = ref $box;
+  my $adj = $box;
+  if($type eq 'ARRAY'){
+    $adj = [map { adjustBoxColor_rec($fg, $bg, $cache, $_); } @$box]; }
+  elsif($type eq 'HASH'){
+    $adj = {map { $_=>adjustBoxColor_rec($fg, $bg, $cache, $$box{$_}); } keys %$box}; }
+  # NASTY access to internal structure; but worth a whole API for this one hack???
+  elsif ($type =~ /^LaTeXML::Core::(?:Box|List|Whatsit|Alignment)$/) { # blessed hashes!
+    $$cache{$box} = $adj = bless {%$box}, $type; # Cache BEFORE recursion!
+    map { $$adj{$_} = adjustBoxColor_rec($fg, $bg, $cache, $$box{$_}); } keys %$box; }
+  elsif($type eq 'LaTeXML::Common::Font'){
```

Comment:
A Font shouldn't make it as an argument to a box-oriented subroutine, surely? This makes porting this to typed languages rather painful, but it is also rather confusing to think about. Why is a font object passed in as `$box`?

Reply:
You're right, bad naming. It's oriented to (digested) objects, not just boxes. Fixed.

Reply:
Is a Font object a Digested object?

I do not understand the design here.

Reply:
A Font object is *inside* of a Digested object, and it is exactly the Font objects inside of the digested objects that we're adjusting.